### PR TITLE
Blog: pytest with Requests a useful automation alternative to Postman in a Python world

### PR DESCRIPTION
**Pull request by mwalker-scottlogic**
https://mwalker-scottlogic.github.io/blog/2025/03/24/pytest-with-requests-a-useful-automation-alternative-to-postman-in-a-python-world.html

Have you (please tick each box to show completion):

 - [x] Added your blog post to a single category?
 - [x] Added a brief summary for your post? Summaries should be roughly two sentences in length and give potential readers a good idea of the contents of your post.
 - [x] Checked that the build passes?
 - [x] Checked your spelling (you can use `npm install` followed by `npx mdspell "**/{FILE_NAME}.md" --en-gb -a -n -x -t` if that's your thing)
 - [x] Ensured that your author profile contains a profile image, and a brief description of yourself? (make it more interesting than just your job title!)
 - [x] Optimised any images in your post? They should be less than 100KBytes as a general guide.

Posts are reviewed / approved by your Regional Tech Lead.

### CODE REVIEW

File: _posts/2025-03-24-pytest-with-requests-a-useful-automation-alternative-to-postman-in-a-python-world.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+*A section of the summary table, showing Riyadh with the highest divergence between forecast and measured AQI data in the specified time range, and data for one of the pollutants*
+
+
+![Graph for pollutant PM2.5 comparing forecasted and actual measurements]({{ site.baseurl }}/mwalkerrose/assets/pollutant-graph.png)
+
+*One of the pollutant (PM2.5) graphs for Riyadh showing the forecast reading with a black line, and the measured data from a specific measurement station in a red line*
+
+
+The rest of this article will contain information about testing the backend of the application, however you can read [this blog](https://blog.scottlogic.com/2025/02/12/playwright-visual-testing.html) to find out more information on UI testing the visual outputs.
```
Comment (jstrong-scottlogic) [inline _posts/2025-03-24-pytest-with-requests-a-useful-automation-alternative-to-postman-in-a-python-world.md]: \* UI-testing

File: _posts/2025-03-24-pytest-with-requests-a-useful-automation-alternative-to-postman-in-a-python-world.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+
+We identified that it would be useful to have our backend integration tests written in the same language as the application, with a key benefit being to facilitate collaboration between developers and testers. If a problem were to arise, having the test code written in Python removes the need for significant context-switching when working together to diagnose and resolve the issue. Similarly for approving pull requests. It also allows the test code to live within the same repository as the source code which would be a benefit as all of the project and documentation could be simply managed through the same repository.
+
+While researching if there was a Python library similar to REST Assured (the framework for testing Java APIs) that we could use for our API tests, we started looking at pytest with the Requests library. The pytest framework was already in the project, being used by the developers to write their unit tests for the backend Python code. However, its use case is not limited to unit testing. Requests is a library that allows HTTP requests to be made from within a pytest test thus extending the pytest functionality and facilitating API testing.
```
Comment (jstrong-scottlogic) [inline _posts/2025-03-24-pytest-with-requests-a-useful-automation-alternative-to-postman-in-a-python-world.md]: \* , thus extending

File: _posts/2025-03-24-pytest-with-requests-a-useful-automation-alternative-to-postman-in-a-python-world.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+An important part of API testing is verifying different parameters work as expected when passed in specific combinations or undertaking boundary value analysis. Both activities can produce a lot of test cases if each test case is quite atomic in its definition and approach. Alternatively with parameterization you can create tests that are more flexible by passing in different arguments in the test setup. These arguments could be, for example, test data and expected response from the API given the data provided. pytest is well suited for this kind of data driven testing due to the inbuilt parameterisation decorator.
```
Comment (jstrong-scottlogic) [inline _posts/2025-03-24-pytest-with-requests-a-useful-automation-alternative-to-postman-in-a-python-world.md]: \* parametrisation, you can create
\* test data and the expected response
\* inbuilt parametrisation decorator

File: _posts/2025-03-24-pytest-with-requests-a-useful-automation-alternative-to-postman-in-a-python-world.md [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+
+## Let’s see how this works.
```
Comment (jstrong-scottlogic) [inline _posts/2025-03-24-pytest-with-requests-a-useful-automation-alternative-to-postman-in-a-python-world.md]: Would probably remove the full stop from this subtitle, or change to 'How it works', for example.

File: _posts/2025-03-24-pytest-with-requests-a-useful-automation-alternative-to-postman-in-a-python-world.md [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+
+In the below simple use case is an example parameterized pytest test using the Python module Requests to make a GET request to the forecast endpoint. The request parameters are passed in as a dictionary ‘payload’. The `location_name parameter` is optional for this endpoint, however `base_time`, `valid_time_from`, `valid_time_to` and `location_type` are required. The test asserts the response has a status code ‘200’ when all required parameters are present, with or without the optional parameter, indicating the request was a success.
```
Comment (jstrong-scottlogic) [inline _posts/2025-03-24-pytest-with-requests-a-useful-automation-alternative-to-postman-in-a-python-world.md]: \* parametrised
\* `location_name` parameter
\* `valid_time_to`, and `location_type` - this is a preference so feel free to ignore.
I don't think 200 needs to be within speech marks or apostrophes.

**Review by jstrong-scottlogic — COMMENTED** [review]
A very enjoyable and informative blog. 😄

**Review by jstrong-scottlogic — APPROVED** [review]

**Review by chriswilty — APPROVED** [review]
